Migration to Vue 3

Carifio24 opened the pull request:
So this is a big one. This PR, in one fell swoop, migrates the Vue pieces of the engine from Vue 2 -> Vue 3 and moves the state management system from Vuex -> Pinia (which is apparently [really just Vuex 5](https://twitter.com/youyuxi/status/1463429442076745730?lang=en)).

## Vue 3

The big change from a structural point of view is that `vue-class-component`, which we currently use to create our components as classes, doesn't support Vue 3. Instead, we now create components using Vue's `defineComponent` helper. This has some implications for state management that I'll mention later. This did force some reorganization of components, as we're now obliged to group all of the methods, computed properties, watchers, etc. together. We're still using what Vue calls the Options API - they seem to recommend using the Composition API, but that would require thought on how to organize our logic into composables.

The global `Vue` API is no longer accessible in Vue 3. Improvements to their reactivity system means that we can just assign to properties and use `delete` rather than `Vue.set` and `Vue.delete`. The biggest implication here is that we can't attach the engine to the global `Vue` object as `Vue.$wwt`, like we currently do. To get around this, I added a `$wwt` member to our Pinia instance which serves the same purpose (we never use `Vue.$wwt` outside of the store anyways).

There are some more mundane package upgrades and replacements which were also necessitated by the upgrade. The main ones to note are:
* `vue-slider-component`, which we use for sliders in the UI, supports Vue 3 via its v4 beta release. I haven't noticed any issues with it, just something to be aware of.
* `v-tooltip` has changed its name to [`floating-vue`](https://github.com/Akryum/floating-vue) and now uses Floating UI (was previously PopperJS) to calculate the popup location. This doesn't seem to play well with our app. I tried a few things to get the positioning right, but ended up switching out our popups to use [`vue3-popper`](https://valgeirb.github.io/vue3-popper/).
* `vue-notification` has a version [ported to v3](https://github.com/kyvg/vue3-notification), which is now what's used here.
* Similarly, `bootstrap-vue` has a Vue 3 port [here](https://github.com/cdmoro/bootstrap-vue-3).
* `vue-color` doesn't support Vue 3, but [`vue-color-kit`](https://vuejsexamples.com/a-lightweight-color-picker-for-vue-3/) was a mostly drop-in replacement (with the exception being that I had to set the CSS for the display inside the layer popup).

## Pinia

Along with the move to Vue 3, this PR also moves the state management from Vuex to Pinia. It probably would've been simpler to do these two pieces separately, there were a few issues with keeping Vuex that didn't seem worth the hassle to me.

In particular, the `defineComponent` helper does a lot of automatic TypeScript deduction, which is nice, but doesn't provide a convenient way (as far as I can tell) for allowing the manual type specifications that we've been using for actions, getters, etc. mapped from the Vuex store. The syntax for mapping with Pinia is mostly the same, but the mapped items now have their types deduced automatically, which means we can dispense with the manual typing.

The store has been reorganized a bit - we're obviously no longer using `vuex-module-decorators`, and Pinia doesn't have mutations - you can either mutate state values directly, or do the operation in an action.

## Other stuff

The BrowserStack tests needed only some minor changes to pass, which I think is a good sign. I disabled a new ESLint rule that wants multi-word component names, since many of ours aren't.

Finally, what do we want to do with the `engine-vuex` module? I assume renaming it to `engine-pinia` is the best route, but I wasn't sure exactly what that entailed as far as our `npm` packaging and such goes, so I haven't touched that yet.

Carifio24 commented:
I've added one more piece to this PR, which is to actually use typing in the Nightwatch tests for the research app. Currently these tests are TypeScript in name only - there's a .ts file that gets transpiled, but we `ts-nocheck` the page object and set the research app page in the tests to be `any`, so it's basically TypeScript in name only. The hangup was that Nightwatch automatically uses the information in the page object file to generate the object that actually gets used in the tests, and I'd never taken the time to work out what the typing should be.

Using [this article](https://www.davidmello.com/using-nightwatch-with-typescript/) as a jumping-off point, I've built out some types for our specific research app page. I debated making this its own PR, but was running into some issues with `@types/nightwatch` and older versions of TypeScript, and there are enough changes to the tests in this PR that it felt appropriate to bundle this in. Note that this typing update doesn't make any additional changes to the test logic.

pkgw commented:
Hmm, looks like the CI is finding some bootstrap-vue problem. Hopefully nothing too tricky to straighten out?

Carifio24 commented:
So I fixed the Bootstrap issue and some other minor things related to ESLint and TypeDoc updates. There's just one issue left that's related to `vue-select`, the library we use for our dropdown component for selecting backgrounds, etc.

The issue is that we're using the beta (`4.0.0-beta.5`) release of `vue-select` (for Vue 3 compatibility), but `@types/vue-select` hasn't caught up. That package still pulls in `vue@^2`, and so we get duplicate type conflicts when building the docs.

We don't use the typing for `vue-select` at all except for the one line in `research-app/main.ts` where we import the component. So one easy solution is to just remove `@types/vue-select` until the actual v4 of that package lands. Not my favorite thing, but it gets the job done. Otherwise, I think we'd need to find a new dropdown component to use.

pkgw commented:
Yeah, it seems like a pretty reasonable solution to just omit the typing for now.

Carifio24 commented:
So the CI shows up as having failed, but all the tests are now passing. Our macOS CI build environment is deprecated and is on a scheduled brownout right now, hence the job cancellation.

I fixed up what were some actual issues, mostly related to our popup component and what elements were on top of one another as the screen gets small. A side effect is that some of the `pointer-events` CSS I added should reduce the amount of container space that blocks events in the tools area to essentially nil.

Carifio24 commented:
@pkgw I bumped the CI version to `macos-11` since that's what currently matches `macos-latest`. It looks like the image is having an issue opening Chrome while executing `npm run test` - not sure right off the bat what's going on. My MacBook is on Monterey, but I'll see if I can reproduce something similar.

pkgw commented:
All green!

pkgw reviewed (commented):
This is phenomenal! It's a lot of changes but it seems like all of the pieces tie together nicely. And the overall stat for the diff is -10k lines of code (maybe mostly lockfile? but still), which is always good. I have a few specific comments below but overall really didn't see much that look like an issue or confused me.

In my functional testing, the performance of HiPS catalogs seems to be a lot worse than what I'm used to. It's definitely possible that this is due to a change in my browser or something, but would you be able to check how this feels for you? I think it's pretty unlikely that anything in this PR would affect this, but maaaaybe there's some new Pinia behavior that's causing a bunch of catalog data to be loaded as reactive state, or something? Also, I still never have the opportunity to click on sources in catalogs — I never get the mouse pointer cursor when I hover over where the sources should be. But that has been an issue for me for a while.

Beyond that, I think we should go ahead and rename `engine-vuex` to `engine-pinia`. Associated with this, we should update the docs — `docs/engine` has some introductory material that will need updating, and we should document that there's a historical `engine-vuex` module but that it won't be maintained going forward. `docs/engine/engine-vuex-index.md` will especially need updating, and the toplevel `README.md`, and probably various files sprinkled around. `git grep -i vuex` will be helpful here as well as `git ls-files |grep -i vuex`. Upon release, I'll update the README for the `@wwtelescope/engine-vuex` package on npmjs.com to point to the updated package.

I tried to run through the functionality of all of the various apps and the pywwt-notebooks demos using the new app, and beyond that *possible* issue with HiPS catalogs, I didn't see any behavior changes. Very impressive, great work!

Oh, also the `internalDepVersions` fields in some of the `package.json` files should be updated to reflect new version requirements: if package A requires new APIs from package B, package A's `package.json` should be updated so that the `"B"` entry of `internalDepVersions` is something like `thiscommit:2022-11-08:asasdiasd`, with the last bit being random junk for disambiguation. If you actually have Cranko installed, `cranko show tctag` will print out such a string for you.

Carifio24 commented:
Yeah, I see what you mean about the HiPS catalogs - comparing side-by-side with the current research app, there is a noticeable performance difference. I'll take a look at what is being stored as reactive - Vue 3 is a lot more flexible with object properties and such being reactive, so it's definitely possible that the issue lies there. I didn't know there were any issues with the point selection - that's working fine on my end.

Once those (and the other few issues you had noticed) are ironed out, I'll dig in to the package naming/versioning stuff.

Carifio24 commented on engine-vuex/src/store.ts:
```diff
@@ -459,750 +462,695 @@ function catalogLayerKey(catalog: CatalogLayerInfo): string {
+  
+      // Un-mirror the spreadsheet layer aspect from the reactivity system. Here
+      // we leverage the quasi-hack that the GUID of the spreadsheet layer is set
+      // to the HiPS dataset name (which is most assuredly not in UUIDv4 format).
```
So I updated the function [here](https://github.com/WorldWideTelescope/wwt-webgl-engine/commit/0b97b230886e3619bbfa6f9d284ad2bf2b23b217) when doing the name -> ID update, but it looks like I forgot to remove the comment. I'll do that now.

Carifio24 commented on embed-creator/src/Creator.vue:
```diff
@@ -505,93 +506,101 @@ function escapeHtml(source: string) {
+      }
+    },
+
+    onClipboardSuccess() {
```
Yep...the package we were using was Vue 2-specific. I obviously took this out and meant to come back to it...then just never came back to it. I've updated this to use `vue-clipboard3`, which seems like the only Vue 3 clipboard package with TS support that anyone is using.

Carifio24 commented:
@pkgw Seems like the performance issues were indeed reactivity-based. The issue was actually in the main WWT component - Vue 3's deep reactivity was causing a re-render of that component every time some internal property of the WWTInstance change. I think it was just particularly noticeable when a HiPS catalog was loaded, since their rendering is more complicated.

I got around this by marking the WWTInstance data member as non-reactive. It's certainly working better for me - let me know if this improves things on your end.

pkgw commented:
Yes, that's definitely much better! Glad that didn't take a bunch of code to fix.

Carifio24 commented:
Ok, great! I think all of the other issues are taken care of now as well, so I'll move on to package renaming.

pkgw commented:
I pushed a couple of tiny final tweaks, but I think this is ready to go in!
